CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc

```diff
@@ -369,16 +377,56 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
 
       // Check whether the trajectory has some missing hits
       bool hasMissingHits = false;
-      for (itm = TMeas.begin(); itm != TMeas.end(); itm++) {
+      unsigned int previous_layer = 999;
+      vector<unsigned int> missedLayers;
+
+      for(const auto& itm : TMeas) {
         auto theHit = (*itm).recHit();
+        unsigned int iidd = theHit->geographicalId().rawId();
+        unsigned int layer = ::checkLayer(iidd, tTopo);
+        int missedLayer = int(layer + 1);
+        int diffPreviousLayer = int(layer - previous_layer);
+        if (doMissingHitsRecovery_) {
+          //Layers from TIB + TOB
+          if (diffPreviousLayer == -2 && missedLayer > k_LayersStart && missedLayer < k_LayersAtTOBEnd) {
+            missHitPerLayer[missedLayer] += 1;
+            hasMissingHits = true;
+          }
+          //Layers from TID
+          if (diffPreviousLayer == -2 && (missedLayer == 12 || missedLayer == 13)) {
+            missHitPerLayer[missedLayer] += 1;
+            hasMissingHits = true;
+          }
+          //Layers from TEC
+          if (diffPreviousLayer == -2 && missedLayer > k_LayersAtTIDEnd && missedLayer < k_LayersAtTECEnd) {
+            missHitPerLayer[missedLayer] += 1;
+            hasMissingHits = true;
+          }
+
+          //##### TID Layer 11 (transition TID -> TIB)
+          if ((int(layer) > k_LayersStart && int(layer) <= k_LayersAtTIBEnd) && (previous_layer == 12)) {
+            missHitPerLayer[11] += 1;
+            hasMissingHits = true;
+          }
+
+          //##### TEC Layer 14 (transition TEC -> TOB)
+          if ((int(layer) > k_LayersAtTIBEnd && int(layer) <= k_LayersAtTOBEnd) && (previous_layer == 15)) {
+            missHitPerLayer[14] += 1;
+            hasMissingHits = true;
+          }
+        }
         if (theHit->getType() == TrackingRecHit::Type::missing)
           hasMissingHits = true;
+
+        if (hasMissingHits)
+          missedLayers.push_back(layer);
+        previous_layer = layer;
       }
 
       // Loop on each measurement and take it into consideration
       //--------------------------------------------------------
-
-      for (itm = TMeas.begin(); itm != TMeas.end(); itm++) {
+      unsigned int prev_TKlayers = 0;
+      for(const auto& itm : TMeas) {
```

Comment:
I don't think using the range-based loop here is appropriate.

Reply:
Sorry for the mistake I went too fast ! Not appropriate here indeed

Reply:
Do you know why changing it back causes an additional error ?

Reply:
there error in not here, it's in the above loop.

Reply:
see https://github.com/cms-sw/cmssw/pull/41970#discussion_r1235265411

Reply:
I can't compile anymore because my version is deprecated, is changing

"auto theHit = (*itm).recHit():"
to
"auto theHit = (itm).recHit()" 

solving this issue in your opinion ?

Reply:
yes it is.